cdap-securestore-spi/src/main/java/co/cask/cdap/securestore/spi/SecretsManager.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ *  use this file except in compliance with the License. You may obtain a copy of
+ *  the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations under
+ *  the License.
+ */
+
+package co.cask.cdap.securestore.spi;
+
+import co.cask.cdap.securestore.spi.secret.Secret;
+import co.cask.cdap.securestore.spi.secret.SecretMetadata;
+
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Secrets Manager interface to store secrets securely and retrieve them when needed. Secrets are small sensitive
+ * information such as passwords, database credentials, API keys etc.
+ *
+ * TODO CDAP-14699 Expose dataset through context in initialize method.
+ */
+public interface SecretsManager {
+  /**
+   * Returns the name of the secrets manager.
+   *
+   * @return non-empty name of this secrets manager
+   */
+  String getName();
+
+  /**
+   * Initialize the secrets manager. This method is guaranteed to be called before any other method is called.
+   * It will only be called once for the lifetime of the secrets manager.
+   *
+   * @param context the context that can be used to initialize the secrets manager
+   */
+  void initialize(SecretsManagerContext context) throws Exception;
+
+  /**
+   * Securely stores secret for a given namespace.
+   *
+   * @param namespace the namespace that this secret belongs to
+   * @param name the name of the secret
+   * @param secret the sensitive data that has to be securely stored
+   * @param description user provided description of the secret
+   * @param properties associated properties and metadata
+   * @throws SecretAlreadyExistsException if the secret already exists in the namespace
+   * @throws Exception if unable to store the secret securely
```

Comment:
Allow throwing Exception is super generic. Do you have more specific exception types that you expect the store method will throw?

Reply:
I think we can add `IOException` as well.